src/Core/Entities/SelfHostedOrganizationDetails.cs

```diff
@@ -0,0 +1,143 @@
+        }
+
+        if (license.MaxCollections.HasValue && CollectionCount > license.MaxCollections.Value)
+        {
+            exception = $"Your organization currently has {CollectionCount} collections. " +
+                $"Your new license allows for a maximum of ({license.MaxCollections.Value}) collections. " +
+                "Remove some collections.";
+            return false;
+        }
+
+        if (!license.UseGroups && UseGroups && GroupCount > 1)
+        {
+            exception = $"Your organization currently has {GroupCount} groups. " +
+                $"Your new license does not allow for the use of groups. Remove all groups.";
+            return false;
+        }
+
+        var enabledPolicyCount = Policies.Count(p => p.Enabled);
+        if (!license.UsePolicies && UsePolicies && enabledPolicyCount > 0)
+        {
+            exception = $"Your organization currently has {enabledPolicyCount} enabled " +
+                $"policies. Your new license does not allow for the use of policies. Disable all policies.";
+            return false;
+        }
+
+        if (!license.UseSso && UseSso && SsoConfig is { Enabled: true })
+        {
+            exception = $"Your organization currently has a SSO configuration. " +
+                $"Your new license does not allow for the use of SSO. Disable your SSO configuration.";
+            return false;
+        }
+
+        if (!license.UseKeyConnector && UseKeyConnector && SsoConfig?.Data != null &&
+            SsoConfig.GetData().KeyConnectorEnabled)
+        {
+            exception = $"Your organization currently has Key Connector enabled. " +
+                $"Your new license does not allow for the use of Key Connector. Disable your Key Connector.";
+            return false;
+        }
+
+        if (!license.UseScim && UseScim && ScimConnections != null &&
+            ScimConnections.Any(c => c.GetConfig<ScimConfig>() is { Enabled: true }))
+        {
+            exception = "Your new plan does not allow the SCIM feature. " +
+                "Disable your SCIM configuration.";
+            return false;
+        }
+
+        if (!license.UseCustomPermissions && UseCustomPermissions &&
+            OrganizationUsers.Any(ou => ou.Type == OrganizationUserType.Custom))
+        {
+            exception = "Your new plan does not allow the Custom Permissions feature. " +
+                "Disable your Custom Permissions configuration.";
+            return false;
+        }
+
+        if (!license.UseResetPassword && UseResetPassword &&
+            Policies.Any(p => p.Type == PolicyType.ResetPassword && p.Enabled))
+        {
+            exception = "Your new license does not allow the Password Reset feature. "
+                + "Disable your Password Reset policy.";
+            return false;
+        }
+
+        exception = "";
+        return true;
+    }
+
+    public Organization ToOrganization()
+    {
+        return new Organization
+        {
```

Comment:
Don't love that we have to do this to extract the `Organization` object out of this subclass. But it is necessary, because Dapper assumes a 1:1 mapping between the object and the table, so we can't give it a superset of the `Organization` object.

The other approach would be to have `Organization` as a property on `SelfHostedOrganizationDetails` rather than the base class, that would preserve the object and avoid this mapping boilerplate, but I that seemed less conceptually clear to me. (i.e. the current object is the organization _with_ extra details, why would the organization be a property on that?) Open to other opinions though.

Reply:
I think having `Organization` as a property makes more sense to me because I like having separation from a database model class and business logic elements.
I think this class should also be moved from `Core.Entities` since that folder should contain table models only. `Core.Models.Business` might make more sense, what do you think?

Reply:
> I think having Organization as a property makes more sense to me because I like having separation from a database model class and business logic elements.

Here's the reason I went with the current implementation: the idea of `SelfHostedOrganizationDetails` is that it's a new **view** model, which aggregates information about an organization from multiple sources. I'd compare it to how `OrganizationUserOrganizationDetails` combines details of `Organization` and `OrganizationUser` entities into a single model. In that example, it's a single flat model made up of information from several table entities, rather than just nesting the table entities as properties. See also [Gibson's comment above](https://github.com/bitwarden/server/pull/2408#discussion_r1021644146).

I've tried implementing the nested `Organization` object as you suggest, just to see how it feels, but it's weird that some details about the organization are nested (`selfHostedOrganizationDetails.Organization.Id`) whereas others aren't (`selfHostedOrganizationDetails.Policies`). These belong to the same conceptual object (the org), why are they treated differently? The answer is that they're located in different database tables - but we shouldn't care about our database structure when referencing this model in the code. It's meant to be an abstraction. In that sense, I think it's already separate from the entity class (even though it subclasses it).

So that's all the "for" argument. I still don't like the `ToOrganization` method, where we have to get the `Organization` entity back _out_ of the model. But at the moment I don't think that outweighs the considerations above.

Does the above convince you? ;) I'm interested in your thoughts either way!

> I think this class should also be moved from Core.Entities since that folder should contain table models only. Core.Models.Business might make more sense, what do you think?

Good point! Core.Entities isn't the right place. But what do you think about putting it in Models.Data.Organizations? That's where other view models live, like `OrganizationAbility` and `OrganizationUserOrganizationDetails`. Either namespace would make sense, but I'm leaning towards Models.Data because it's pulled directly from the database.

Reply:
> Does the above convince you? ;) I'm interested in your thoughts either way!
I am convinced! It's a good use case of a view model.

> Good point! Core.Entities isn't the right place. But what do you think about putting it in Models.Data.Organizations? That's where other view models live
You made a good point, there's already a lot of ViewModel classes in that namespace. I'm fine with adding another one but I feel like we should differentiate the Entity classes from the rest. That's out of scope for this so we can think about it some point in the future.